---
output:
  bookdown::html_document2:
    highlight: tango
    number_sections: no

params:
  config: snakemake.params.config
  set_author: snakemake.params.author
  set_email: snakemake.params.email
---

---
title: RNA-Seq and DE analysis
author:
  - Bioinformatics Core Facility, Centre for Molecular Medicine, CEITEC Masaryk University
  - ^1^CEITEC-Central European Institute of Technology, Masaryk University, Kamenice 5, 625 00 Brno, Czech Republic
  - `r params$set_author`^1^ [`r params$set_email`](mailto:`r params$set_email`)
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
```

```{r load-packages, include=FALSE}
library(data.table)
library(jsonlite)
library(flextable)
library(stringr)

knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
set_flextable_defaults(font.family = "Arial", font.size = 12, padding.top = 3, padding.bottom = 3)
```

```{r variables, results='asis', echo=FALSE}
config_json <- as.data.table(fromJSON(txt = params$config))
config_json[,condition := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$condition)]
  config_json[,replicate       := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$replicate)]
  config_json[,full_name := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$sample_name)]

comparison <- unique(config_json$comparison)
biotype <- unique(config_json$biotype_list)

if(config_json$RSEM[1] == TRUE){
  var.rsem <- 1
  DE <- "DE_RSEM"
}else{
  var.rsem <- 0
  DE <- "DE_feature_count"
}

`%!in%` <- Negate(`%in%`)

mRNA_DE_FOLDER_PATH <- paste0("../results/",DE,"/")
INPUTFILES <- paste0("../qc_reports/")
MQC_DATA <- paste0(INPUTFILES,"all_samples/","multiqc_data/")

# set of colors for plot to look like multiqc
highchart.v4 <- c("#7cb5ec", "#434348", "#90ed7d", "#f7a35c", "#8085e9", "#f15c80", "#e4d354", "#2b908f", "#f45b5b", "#91e8e1")
rseqc.palette <- c("#7cb5ec","#434348","#90ed7d","#f7a35c","#8085e9","#f15c80","#e4d354","#2b908f","#f45b5b","#91e8e1","#cccccc","#7f0000")
biobloom.palette <- c("#7cb5ec","#434348","#90ed7d","#f7a35c","#8085e9","#f15c80","#e4d354","#2b908f","#f45b5b","#91e8e1","#148E0B","#FDE74C","#7f0000","#cccccc")
star.palette <- c("#437bb1","#7cb5ec","#f7a35c","#e63491","#b1084c","#7f0000")
```

# Acknowledgement

- We request co-authorship (and will assist with manuscript preparation) if we have developed novel tools, algorithms, or pipelines, participated in experiment design planning, or have contributed to a biological question addressed in a manuscript. We ask that, at a minimum, you acknowledge us in a publication to which we have contributed routine analysis without further support and discussions, data management or conversion, or data submission services.
- Core Facility Bioinformatics of CEITEC Masaryk University is gratefully acknowledged for the obtaining of the scientific data presented in this paper.

# Description
```{r, results='asis', echo=FALSE}
species <- config_json$species[1]
reference <- config_json$reference[1]
num_conditions <- length(unique(config_json$condition))
num_replicat <- length(unique(config_json$replicate))
rna_type <- config_json$RNAseq_type[1]
strandness <- config_json$strandness[1]
umi <- config_json$UMI[1]

paired <- ifelse(config_json$is_paired == TRUE, "paired-end", "single-end")

tabl_sample <- fread(paste0(MQC_DATA,"multiqc_trimmomatic.txt"), header=T)

```
- The main goal of the experiment is to perform Differential Expression analysis.
- In general, we have `r length(config_json[,full_name])` `r species` samples. We have `r num_conditions` conditions, those conditions are sequenced in `r num_replicat` replicates, which is sufficient to obtain statistical significance.
- Conditions to compare are:
`r border_remove(theme_booktabs(delete_part(qflextable(data = as.data.frame(unique(config_json$comparison))),part = "header")))`

`r if(rna_type == "classic_rev"){paste0("- Samples were sequenced using ",paired," 75 bp reads; NEBNext Ultra II Directional RNA Library Prep Kit for Illumina + polyA selection.")}`
`r if(rna_type == "quant_fwd"){"- Samples were sequenced using Lexogen Quantseq FWD kit, therefore a lot of reads in UTR regions are expected as well as huge 3â€™ sequencing bias. These data can only be used for differential expression analysis."}`
`r if("No" %!in% umi){"- Samples contained 6bp long UMIs used for detection of PCR duplicates in library."}`

## Agreed task(s)
1. General samples QC.
2. RNA-Seq analysis.
3. DE analysis.

## Samples
- Brief sample description and number of raw and preprocessed reads are summarized in Table \@ref(tab:tabl-sample).

```{r tab.id="tabl-sample", tab.cap="Sample description - condition assignment and number of raw and preprocessed reads.",  results='asis', echo=FALSE, ft.split=F}
tabl_sample <- merge(config_json[, .(Sample = full_name, condition)], tabl_sample, by="Sample")
setorder(tabl_sample, condition, Sample)

fit_to_width(autofit(qflextable(tabl_sample[, .(Sample, condition, "raw reads" = input_reads, "preprocessed reads" = surviving)])),max_width = 7.1)
```



