---
output:
  bookdown::html_document2:
    highlight: tango
    number_sections: false
    global_numbering: true
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: true

params:
  config: snakemake.params.config
  set_author: snakemake.params.author
  set_email: snakemake.params.email
---

---
title: RNA-Seq and DE analysis
author:
  - Bioinformatics Core Facility, Centre for Molecular Medicine, CEITEC Masaryk University
  - ^1^CEITEC-Central European Institute of Technology, Masaryk University, Kamenice 5, 625 00 Brno, Czech Republic
  - `r params$set_author`^1^ [`r params$set_email`](mailto:`r params$set_email`)
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
```

```{r load-packages, include=FALSE}
library(data.table)
library(jsonlite)
library(flextable)
library(stringr)
library(ggplot2)
library(TidyMultiqc)
library(tidyr)
library(officer)

knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
set_flextable_defaults(font.family = "Arial", font.size = 12, padding.top = 3, padding.bottom = 3)
```

```{r variables, results='asis', echo=FALSE}
config_json <- as.data.table(fromJSON(txt = params$config))
config_json[,condition := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$condition)]
  config_json[,replicate := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$replicate)]
  config_json[,full_name := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$sample_name)]

comparison <- unlist(strsplit(unique(config_json$comparison),split="\\|"))
biotype <- unlist(strsplit(unique(config_json$biotype_list),split="\\|"))

if(config_json$RSEM[1] == TRUE){
  var.rsem <- 1
  DE <- "DE_RSEM"
}else{
  var.rsem <- 0
  DE <- "DE_feature_count"
}

`%!in%` <- Negate(`%in%`)

mRNA_DE_FOLDER_PATH <- paste0("../results/",DE,"/")
INPUTFILES <- paste0("../qc_reports/")
MQC_DATA <- paste0(INPUTFILES,"all_samples/","multiqc_data/")

# set of colors for plot to look like multiqc
highchart.v4 <- c("#7cb5ec", "#434348", "#90ed7d", "#f7a35c", "#8085e9", "#f15c80", "#e4d354", "#2b908f", "#f45b5b", "#91e8e1")
rseqc.palette <- c("#7cb5ec","#434348","#90ed7d","#f7a35c","#8085e9","#f15c80","#e4d354","#2b908f","#f45b5b","#91e8e1","#cccccc","#7f0000")
biobloom.palette <- c("#7cb5ec","#434348","#90ed7d","#f7a35c","#8085e9","#f15c80","#e4d354","#2b908f","#f45b5b","#91e8e1","#148E0B","#FDE74C","#7f0000","#cccccc")
star.palette <- c("#437bb1","#7cb5ec","#f7a35c","#e63491","#b1084c","#7f0000")
```

# Acknowledgement

- We request co-authorship (and will assist with manuscript preparation) if we have developed novel tools, algorithms, or pipelines, participated in experiment design planning, or have contributed to a biological question addressed in a manuscript. We ask that, at a minimum, you acknowledge us in a publication to which we have contributed routine analysis without further support and discussions, data management or conversion, or data submission services.
- Core Facility Bioinformatics of CEITEC Masaryk University is gratefully acknowledged for the obtaining of the scientific data presented in this paper.

# Description
```{r samples, results='asis', echo=FALSE}
species <- config_json$species[1]
reference <- config_json$reference[1]
num_samples <- length(config_json[,full_name])
num_conditions <- length(unique(config_json$condition))
num_replicat <- length(unique(config_json$replicate))
rna_type <- config_json$RNAseq_type[1]
if(config_json$strandness[1] == "fwd"){
  strandness <- "Forward"
}else if(config_json$strandness[1] == "rev"){
  strandness <- "Reverse"
}else if(config_json$strandness[1] == "unstr"){
  strandness <- "Unstranded"
}
umi <- config_json$UMI[1]

paired <- ifelse(config_json$is_paired[1] == TRUE, "paired-end", "single-end")

tabl_sample <- fread(paste0(MQC_DATA,"multiqc_trimmomatic.txt"), header=T)
tabl_sample[, Sample := gsub("_R.","",Sample)]
if(is.null(tabl_sample$input_reads)==T){tabl_sample[,input_reads:=input_read_pairs]}
tabl_sample <- merge(config_json[, .(Sample = full_name, condition)], tabl_sample, by="Sample")
setorder(tabl_sample, condition, Sample)

# Height for sample plots
# default is good for few samples and Read coverage plot
default_height <- 5
# custom is needed if several samples >= 50 ? -> 10
custom_height1 <- ifelse(num_samples < 40, default_height, 10)
# another custom if the first one is not enough (rRNA plot can need higher value because of pair-end doubling the amount of samples) -> 15
custom_height2 <- ifelse(paired == "paired-end" & num_samples >= 40, 15, custom_height1)
```
- The main goal of the experiment is to perform Differential Expression analysis.
- In general, we have `r num_samples` `r species` samples. We have `r num_conditions` conditions, those conditions are sequenced in `r num_replicat` replicates, which is sufficient to obtain statistical significance.
- Conditions to compare are:
`r border_remove(theme_booktabs(delete_part(qflextable(data = as.data.frame(unique(comparison))),part = "header")))`

`r if(rna_type == "classic_rev"){paste0("- Samples were sequenced using ",paired," 75 bp reads; NEBNext Ultra II Directional RNA Library Prep Kit for Illumina + polyA selection.")}`
`r if(rna_type == "quant_fwd"){"- Samples were sequenced using Lexogen Quantseq FWD kit, therefore a lot of reads in UTR regions are expected as well as huge 3’ sequencing bias. These data can only be used for differential expression analysis."}`
`r if("No" %!in% umi){"- Samples contained 6bp long UMIs used for detection of PCR duplicates in library."}`

## Agreed task(s)
1. General samples QC.
2. RNA-Seq analysis.
3. DE analysis.

## Samples
- Brief sample description and number of raw and preprocessed reads are summarized in Table \@ref(tab:tabl-sample).

```{r tab.id="tabl-sample", tab.cap="Sample description - condition assignment and number of raw and preprocessed reads.",  results='asis', echo=FALSE, ft.split=F}
fit_to_width(autofit(qflextable(tabl_sample[, .(Sample, condition, "raw reads" = input_reads, "preprocessed reads" = surviving)])),max_width = 9.1)
```

## Analysis
```{r rseqc, echo=FALSE}
rseqc.infer.exp <- fread(paste0(MQC_DATA,"multiqc_rseqc_infer_experiment.txt"))
rseqc.infer.exp <- rseqc.infer.exp[,Sample := gsub(".RSeQC.*","",Sample)]
rseqc.infer.exp <- merge(config_json[, .(Sample = full_name, condition)], rseqc.infer.exp, by="Sample")
setorder(rseqc.infer.exp, condition, Sample)

if(paired=="single-end"){
  rseqc.sense <- round(mean(100*rseqc.infer.exp$se_sense),0)
  rseqc.antisense <- round(mean(100*rseqc.infer.exp$se_antisense),0)
  rseqc.failed <- round(mean(100*rseqc.infer.exp$failed),0)
}else{
  rseqc.sense <- round(mean(100*rseqc.infer.exp$pe_sense),0)
  rseqc.antisense <- round(mean(100*rseqc.infer.exp$pe_antisense),0)
  rseqc.failed <- round(mean(100*rseqc.infer.exp$failed),0)
}

rseqc.infer.exp <- melt(rseqc.infer.exp, id.vars = c("Sample","condition"))
rseqc.infer.exp <- rseqc.infer.exp[, value := 100*value]
pct_antisense <- rseqc.infer.exp[variable %like% "antisense", .(mean(value))]

if(pct_antisense >= 70){strandness.found <- "Reverse"}
if(pct_antisense < 30){strandness.found <- "Forward"}
if(pct_antisense >= 30 & pct_antisense < 70){strandness.found <- "Unstranded"}

plot.rseqc.infer.exp <- ggplot(rseqc.infer.exp, aes(Sample, value, fill = variable)) +
  geom_bar(stat = "identity" , width = 0.8, position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = rseqc.palette, labels = c("Sense", "Antisense", "Undetermined"), name = "") +
  coord_flip() +
  #scale_y_continuous(labels = scales::percent, name = "% Tags", breaks = c(0,0.2,0.4,0.6,0.8,1)) +
  scale_y_continuous(name = "% Tags", breaks = seq(0,100,10)) +
  xlab("") +
  scale_x_discrete(limits=rev) +
  theme_bw() +
  theme(legend.position="bottom") +
  theme(legend.key.size = unit(0.6,"line")) +
  theme(plot.title = element_text(face="bold")) +
  ggtitle("RSeQC: Infer experiment")

```
- Raw reads were quality checked (**FastQC, MultiQC, minion, swan**), preprocessed (**Trimmomatic, FastQC, MultiQC**) and mapped (**STAR, Samtools, MultiQC**) to the reference genome with gene annotation (**genome version: `r reference`**).
- Mapped reads were counted and summarized to genes (**featureCounts**).
  - Only uniquely mapped and uniquely assigned reads were counted.
`r if(var.rsem==1){"- Mapped reads were counted and summarized to genes (**RSEM**)."}`
`r if(var.rsem==1){"  - Uniquely mapped as well as multimapped reads were counted."}`
- The library preparation kit strand-specificity is: **`r tolower(strandness)`**, and only reads in this strand should be considered.
- Following checks were performed to estimate the overall sample quality:
  - rRNA content estimate (**fastq_screen**).
  - Read duplication rate (**dupRadar, Picard tools**).
  - Sequenced (targeted) regions (**RSeQC, Picard tools**).
  - 5’/3’ coverage bias (**Picard tools**).
  - Expressed gene biotypes (**featureCounts**).
  - Library strandness (**RSeQC**).
  - Other quality checks are available upon request.
- Complete settings, used commands, tool versions and methodology part for a publication can be provided upon request.

## Results
### General samples QC
- The sequencing depth varies from ~`r round(tabl_sample[input_reads == min(input_reads), input_reads]/1000000)`M of reads for sample `r tabl_sample[input_reads == min(input_reads), Sample]` to ~`r round(tabl_sample[input_reads == max(input_reads), input_reads]/1000000)`M of reads for sample `r tabl_sample[input_reads == max(input_reads), Sample]`.

```{r adaptors, results='asis', echo=FALSE}

if(file.exists(paste0(MQC_DATA,"mqc_fastqc_adapter_content_plot_1.txt"))){
 adaptor_max <- fread(paste0(MQC_DATA,"mqc_fastqc_adapter_content_plot_1.txt"), header=T)
 adaptor_max <- melt(adaptor_max, id.vars = c("Sample"), value.name = "plot.adaptor.max")
}else{
 #adaptor_max <- load_multiqc(paste0(MQC_DATA,"multiqc_data.json"), sections = c('plots'), plot_opts = list(fastqc_adapter_content_plot= list(extractor = extract_ignore_x, summary = list(max = max), prefix = "adaptor"))) # tidymultiqc v0.1.0
 adaptor_plot <- load_multiqc(paste0(MQC_DATA,"multiqc_data.json"), sections = 'plot', plots = 'fastqc_adapter_content_plot')
 adaptor_max <- adaptor_plot %>% tidyr::unnest(cols = plot.fastqc_adapter_content_plot) %>%
         dplyr::group_by(metadata.sample_id) %>%
         dplyr::summarise(plot.adaptor.max = max(y))
}

if("plot.adaptor.max" %in% colnames(adaptor_max) == TRUE){
 max_adaptor <- round(max(adaptor_max$plot.adaptor.max, na.rm = TRUE),0)
}else{
 max_adaptor <- 0.1 # in some cases multiqc is not reporting the plot if "No samples found with any adapter contamination > 0.1%"
}

```

- Initial quality check for analysis did not show presence of adapters (<`r max_adaptor`%), no adapter trimming was performed
- Other initial quality checks look OK

### Alignment and splices
```{r star, results='asis', echo=FALSE}
multiqc_star <- fread(paste0(MQC_DATA,"multiqc_star.txt"))
multiqc_star <- merge(config_json[, .(Sample = full_name, condition)], multiqc_star, by="Sample")

tabl_mapread <- multiqc_star[, .(Sample, condition, uniquely_mapped, uniquely_mapped_percent, multimapped, multimapped_percent)]
setorder(tabl_mapread, condition, Sample)

uni.map.mean.perc <- round(mean(tabl_mapread[,uniquely_mapped_percent]), 0)
uni.map.min <- round(min(tabl_mapread[,uniquely_mapped]/1E6))
uni.map.max <- round(max(tabl_mapread[,uniquely_mapped]/1E6))

col.perc <- names(multiqc_star) %like% "percent"
multiqc_star_perc <- cbind(multiqc_star[,1], multiqc_star[, ..col.perc])
multiqc_star_perc <- melt(multiqc_star_perc, id.vars = "Sample")
multiqc_star_perc <- multiqc_star_perc[, variable := gsub("_percent","",variable)]
multiqc_star_perc <- multiqc_star_perc[, variable := factor(variable, levels = c("uniquely_mapped","multimapped","multimapped_toomany","unmapped_mismatches","unmapped_tooshort","unmapped_other"))]
star.labcol <- multiqc_star_perc[, .(variable = unique(variable) , Colour = star.palette, Label = c("Uniquely mapped", "Mapped to multiple loci", "Mapped to too many loci","Unmapped: too many missmatches","Unmapped: too short","Unmapped: other"))]

star.name.not0 <- multiqc_star_perc[, .(sum = sum(value)), by = variable][sum > 0,]$variable
multiqc_star_perc <- multiqc_star_perc[variable %in% star.name.not0,]
star.labcol <- star.labcol[variable %in% star.name.not0,]


plot.multiqc_star_perc <- ggplot(multiqc_star_perc, aes(Sample, value, fill = variable)) +
  geom_bar(stat = "identity" , width = 0.8, position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = star.labcol$Colour, labels = star.labcol$Label, name = "") +
  coord_flip() +
  scale_y_continuous(name = "Percentages", breaks = seq(0,100,10)) +
  xlab("") +
  scale_x_discrete(limits=rev) +
  theme_bw() +
  theme(legend.position="bottom") +
  theme(legend.key.size = unit(0.6,"line")) +
  guides(fill=guide_legend(nrow=3, byrow=TRUE)) +
  theme(plot.title = element_text(face="bold")) +
  ggtitle("STAR: Alignment Scores")

```

- The single end reads were aligned to `r species` reference genome with the help of the gene annotation (`r reference`) using STAR aligner.
- The number and percentage of uniquely- and multi-mapped reads are pretty good for all samples ~`r uni.map.mean.perc`% of uniquely mapped and we are getting approximately the same values across all samples, same applies for number of mapped reads, which varies between ~`r uni.map.min`M and ~`r uni.map.max`M (Table \@ref(tab:tabl-mapread), Figure \@ref(fig:star-map)).

```{r tab.id="tabl-mapread", tab.cap="Number and percentage of uniquely- and multi-mapped reads", results='asis', echo=FALSE, out.width="80%", ft.split=F}
fit_to_width(autofit(qflextable(tabl_mapread[,.(Sample, condition, "uniquely mapped" = uniquely_mapped, "uniquely mapped (%)" = uniquely_mapped_percent, multimapped, "multimapped (%)" = multimapped_percent)])),max_width = 9.1)
```

```{r star-map, results='asis', echo=FALSE, fig.cap="Percentages and distribution of different mapping types. For more details please see the STAR aligner manual.", fig.align="center", out.width="90%", fig.show="hold", dev="svg", fig.width=7, fig.height = custom_height1}
plot.multiqc_star_perc
```

### rRNA contamination
```{r rrna, results='asis', echo=FALSE}
fastq.screen.files <- fread(paste0(MQC_DATA,"multiqc_fastq_screen.txt"))
#fastq.screen.files <- merge(config_json[, .(Sample = full_name, condition)], fastq.screen.files[, Sample := gsub("_screen$","",Sample)], by="Sample")
fastq.screen.files <- fastq.screen.files[, Sample := gsub("_screen$","",Sample)]

rrna <- fastq.screen.files[, .(Sample, rRNA.perc=get("rRNA percentage"))]

rrna.p <- ggplot(rrna, aes(Sample, rRNA.perc)) +
  geom_bar(stat = "identity", width = 0.8, fill = "#7f0000") +
  theme_bw() +
  coord_flip() +
  ylab("rRNA content (%)") +
  xlab("") +
  scale_x_discrete(limits=rev) +
  theme(plot.title = element_text(face="bold")) +
  ggtitle("rRNA Content")

rrna.max <- round(max(rrna[,rRNA.perc]))
```

- Samples had a percentage of rRNA contamination up to `r rrna.max`%
  - For polyA selection, we would expect ~ 1%, while for rRNA depletion it could be up to 10%.
  - This can be due to less effective polyA selection procedure, but overall we still have a lot of usable reads and information to perform Differential expression analysis.
  - The percentages are summarized in Figure \@ref(fig:fastq-screen-rrna).

```{r fastq-screen-rrna, results='asis', echo=FALSE, fig.cap="rRNA estimate.", fig.align="center", out.width="90%", fig.show="hold", dev="svg", fig.width=7, fig.height = custom_height2}
rrna.p
```

### Mapped regions
```{r metrics, echo=FALSE}
picard.metric <- fread(paste0(MQC_DATA,"multiqc_picard_RnaSeqMetrics.txt"))
picard.metric <- merge(config_json[, .(Sample = full_name, condition)], picard.metric, by="Sample")

rseqc.metric <- fread(paste0(MQC_DATA,"multiqc_rseqc_read_distribution.txt"))
rseqc.metric <- merge(config_json[, .(Sample = full_name, condition)], rseqc.metric[, Sample:=gsub(".RSeQC.*","",Sample)], by="Sample")

rseqc.pct <- names(rseqc.metric)[names(rseqc.metric) %like% "tag_pct"]
rseqc.metric.tbl <- cbind(rseqc.metric[,1], rseqc.metric[, ..rseqc.pct])
#rseqc.metric.tbl <- rseqc.metric.tbl[, other_intergenic_tag_pct := 100-rowSums(.SD), .SDcols = 2:11]
rseqc.metric.tbl <- rseqc.metric.tbl[, `:=` (true_tss_up_5kb_tag_pct = tss_up_5kb_tag_pct - tss_up_1kb_tag_pct,
                                             true_tss_up_10kb_tag_pct = tss_up_10kb_tag_pct - tss_up_5kb_tag_pct,
                                             true_tes_down_5kb_tag_pct = tes_down_5kb_tag_pct - tes_down_1kb_tag_pct,
                                             true_tes_down_10kb_tag_pct = tes_down_10kb_tag_pct - tes_down_5kb_tag_pct)]
rseqc.metric.tbl <- rseqc.metric.tbl[, `:=` (tss_up_5kb_tag_pct = NULL,
                                             tss_up_10kb_tag_pct = NULL,
                                             tes_down_5kb_tag_pct = NULL,
                                             tes_down_10kb_tag_pct = NULL)]

rseqc.metric <- rseqc.metric[, "mrna_tag_pct" := (get("cds_exons_tag_pct") + get("5_utr_exons_tag_pct") + get("3_utr_exons_tag_pct"))]

picard.intron <- round(mean(picard.metric$PCT_INTRONIC_BASES),0)
rseqc.intron <- round(mean(rseqc.metric$introns_tag_pct),0)
picard.mrna <- round(mean(picard.metric$PCT_MRNA_BASES),0)
rseqc.mrna <- round(mean(rseqc.metric$mrna_tag_pct),0)

if(picard.intron >= 40){
  intron.count <- "very high"
}else if(picard.intron >= 20 & picard.intron < 40){
  intron.count <- "high"
}else if(picard.intron >= 5 & picard.intron < 20){
  intron.count <- "low"
}else if(picard.intron < 5){
  intron.count <- "very low"
}

picard.metric.tbl <- picard.metric[, .(Sample, Coding = CODING_BASES/PF_BASES, UTR = UTR_BASES/PF_BASES, Intronic = INTRONIC_BASES/PF_BASES, Intergenic = INTERGENIC_BASES/PF_BASES, not_aligned = PF_NOT_ALIGNED_BASES/PF_BASES)]
picard.metric.tbl <- melt(picard.metric.tbl, id.vars = "Sample")
picard.metric.tbl <- picard.metric.tbl[, value := 100*value]

plot.picard.metric <- ggplot(picard.metric.tbl, aes(Sample, value, fill = factor(variable, levels = c("Coding","UTR","Intronic","Intergenic","not_aligned")))) +
  geom_bar(stat = "identity" , width = 0.8, position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = rseqc.palette, labels = c("Coding","UTR","Intronic","Intergenic","PF not aligned"), name = "") +
  coord_flip() +
  scale_y_continuous(name = "Percentages", breaks = seq(0,100,10)) +
  xlab("") +
  scale_x_discrete(limits=rev) +
  theme_bw() +
  theme(legend.position="bottom") +
  theme(legend.key.size = unit(0.6,"line")) +
  guides(fill=guide_legend(nrow=1, byrow=TRUE)) +
  theme(plot.title = element_text(face="bold")) +
  ggtitle("Picard: RnaSeqMetrics Read Assignments")

rseqc.metric.tbl <- melt(rseqc.metric.tbl, id.vars = "Sample")
rseqc.metric.tbl <- rseqc.metric.tbl[, `:=` (Sample = gsub(".RSeQC.*","",Sample), variable = gsub("_tag_pct","",variable))]

plot.rseqc.metric <- ggplot(rseqc.metric.tbl, aes(Sample, value, fill = factor(variable, levels = c("cds_exons","5_utr_exons","3_utr_exons","introns","tss_up_1kb","true_tss_up_5kb","true_tss_up_10kb","tes_down_1kb","true_tes_down_5kb","true_tes_down_10kb","other_intergenic")))) +
  geom_bar(stat = "identity" , width = 0.8, position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = rseqc.palette, labels = c("CDS_Exons","5'UTR_Exons","3'UTR_Exons","Introns","TSS_up_1kb","TSS_up_5kb","TSS_up_10kb","TES_down_1kb","TES_down_5kb","TES_down_10kb","Other_intergenic"), name = "") +
  coord_flip() +
  scale_y_continuous(name = "Percentages", breaks = seq(0,100,10)) +
  xlab("") +
  scale_x_discrete(limits=rev) +
  theme_bw() +
  theme(legend.position="bottom") +
  theme(legend.key.size = unit(0.6,"line")) +
  guides(fill=guide_legend(nrow=3, byrow=TRUE)) +
  theme(plot.title = element_text(face="bold")) +
  ggtitle("RSeQC: Read Distribution")

```

- Evaluation of the mapped genomic regions is very comparable
- All samples had approximately a low content of intronic reads ~`r picard.intron`% counted over bases as well as ~`r rseqc.intron`% counted over reads which is expected for polyA selection library protocols
- All samples had majority of reads ~`r picard.mrna`% counted over bases as well as ~`r rseqc.mrna`% counted over reads mapped within coding regions (`r paste0("Figures \\@ref(fig:picard-map-loc) and \\@ref(fig:rseqc-map-loc)")`).
- All samples are good quality and can be safely used for Differential expression analysis.

```{r picard-map-loc, results='asis', echo=FALSE, fig.cap="Mapped locations (Picard; in bases).", fig.align="center", out.width="90%", fig.show="hold", dev="svg", fig.width=7, fig.height = custom_height1}
plot.picard.metric
```

```{r rseqc-map-loc, results='asis', echo=FALSE, fig.cap="Mapped locations (RSeQC; in reads).", fig.align="center", out.width="90%", fig.show="hold", dev="svg", fig.width=7, fig.height = custom_height1}
plot.rseqc.metric
```

### Read coverage distribution
- The gene body coverage is very similar in all the samples.
`r if(rna_type == "classic_rev"){"  - Coverage across the gene is very even and shows very good coverage across whole genes "}else if(rna_type == "quant_fwd"){"  - Coverage across the gene is very good and shows the characteristic accumulation in 3'end of the genes in all the samples, which is typical from the library preparation using QuantSeq "}``r paste0("(Figure \\@ref(fig:picard-cov)).")`

```{r picard-cov, results='asis', echo=FALSE, fig.cap="Gene body mapping distribution.", fig.align="center", out.width="90%", fig.show="hold", dev="svg", fig.width=7, fig.height = default_height}
# gene_cov <- load_multiqc(
#   paste0(MQC_DATA,"multiqc_data.json"),
#   sections = "plots",
#   plot_opts = list(picard_rna_coverage = list(
#     extractor = extract_xy,
#     summary = list(
#       df=list
#     )
#   ))
# )
gene_cov<-load_multiqc(paste0(MQC_DATA,"multiqc_data.json"), sections = 'plot', plots = 'picard_rna_coverage')
#gene_cov <- as.data.table(gene_cov %>% unnest(plot.picard_rna_coverage.df))
gene_cov <- as.data.table(gene_cov %>% unnest(plot.picard_rna_coverage))
setnames(gene_cov, "metadata.sample_id","Sample")

gene_cov <- merge(config_json[, .(Sample = full_name)], gene_cov, by="Sample")

coul <- c(rseqc.palette,rseqc.palette)
if(length(unique(gene_cov$Sample)) > length(coul)){
  coul <- colorRampPalette(coul)(length(unique(gene_cov$Sample)))
}

plot.gene.cov <- ggplot(gene_cov, aes(x, y, colour = factor(Sample))) +
  geom_line() +
  scale_colour_manual(values = coul, name = "") +
  ylab("Coverage") +
  scale_x_continuous(name = "Percent through gene", breaks = seq(0,100,10)) +
  theme_bw() +
  theme(legend.position="none") +
  theme(plot.title = element_text(face="bold")) +
  ggtitle("Picard: Normalized Gene Coverage")

plot.gene.cov

```


### Strand specificity
- According to the preparation kit reads should be `r if(strandness == "Forward"){"in the forward strand."}else if(strandness == "Reverse"){"in the reverse strand."}else if(strandness == "Unstranded"){"unstranded."}`
- The analysis shows the strand-specificity of each library, but cannot be trusted with full confidence.
  - All the samples have ~`r rseqc.sense`% of the forward-strand specificity, ~`r rseqc.antisense`% of the reverse-strand specificity but ~`r rseqc.failed`% mappings cannot be determined.
    - The latest are usually mappings to intergenic regions, unannotated regions or regions where two genes overlap at the same strand (`r paste0("Figure \\@ref(fig:rseqc-exp)")`).
  - We consider mapping as `r tolower(strandness.found)`, as we got the highest number of suitable reads for DE analysis.

```{r rseqc-exp, results='asis', echo=FALSE, fig.cap="Library strand specificity.", fig.align="center", out.width="90%", fig.show="hold", dev="svg", fig.width=7, fig.height = custom_height1}
plot.rseqc.infer.exp
```

### Read count assignment to genes
```{r fc-rsem, echo=FALSE}
if(var.rsem == 1){
rsem.assign <- fread(paste0(MQC_DATA,"multiqc_rsem.txt"))
rsem.assign <- rsem.assign[, aligned.unique := 100*Unique/Total]
rsem.assign <- rsem.assign[, aligned.multi := 100*Multi/Total]
rsem.assign <- rsem.assign[, filter := 100*Filtered/Total]
rsem.assign <- rsem.assign[, unaligned := 100*Unalignable/Total]

rsem.assign.tabl <- rsem.assign[, .(Sample, aligned.unique, aligned.multi, filter,unaligned)]
rsem.assign.tabl <- melt(rsem.assign.tabl, id.vars = "Sample")

plot.rsem.assign <- ggplot(rsem.assign.tabl, aes(Sample, value, fill = factor(variable, levels = c("aligned.unique","aligned.multi","filter","unaligned")))) +
  geom_bar(stat = "identity" , width = 0.8, position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = star.palette[c(1,4,5,6)], name = "", labels = c("Aligned uniquely to a gene","Aligned to multiple genes","Filtered due to too many alignments","Unalignable reads")) +
  coord_flip() +
  scale_y_continuous(name = "Percentages", breaks = seq(0,100,10)) +
  xlab("") +
  scale_x_discrete(limits=rev) +
  theme_bw() +
  theme(legend.position="bottom") +
  theme(legend.key.size = unit(0.6,"line")) +
  guides(fill=guide_legend(nrow=2, byrow=TRUE)) +
  theme(plot.title = element_text(face="bold")) +
  ggtitle("RSEM: Mapped reads")

rsem.assign.min <- round(min(rsem.assign[,Alignable]/1E6),0)
rsem.assign.max <- round(max(rsem.assign[,Alignable]/1E6),0)
rsem.assign.min.perc <- round(min(rsem.assign[,alignable_percent]),0)
rsem.assign.max.perc <- round(max(rsem.assign[,alignable_percent]),0)
}

fc.assign <- fread(paste0(MQC_DATA,"multiqc_featureCounts.txt"))
fc.assign <- merge(config_json[, .(Sample = full_name, condition)], fc.assign, by="Sample")
setorder(fc.assign, condition, Sample)
fc.assign.min <- round(min(fc.assign[,Assigned]/1E6),0)
fc.assign.max <- round(max(fc.assign[,Assigned]/1E6),0)
fc.assign.min.perc <- round(min(fc.assign[,percent_assigned]),0)
fc.assign.max.perc <- round(max(fc.assign[,percent_assigned]),0)
fc.assign <- fc.assign[, `:=` (Assigned_pct = 100*Assigned/Total,
                               Unassigned_Unmapped_pct = 100*Unassigned_Unmapped/Total,
                               Unassigned_MappingQuality_pct = 100*Unassigned_MappingQuality/Total,
                               Unassigned_Chimera_pct = 100*Unassigned_Chimera/Total,
                               Unassigned_FragmentLength_pct = 100*Unassigned_FragmentLength/Total,
                               Unassigned_Duplicate_pct = 100*Unassigned_Duplicate/Total,
                               Unassigned_MultiMapping_pct = 100*Unassigned_MultiMapping/Total,
                               Unassigned_Secondary_pct = 100*Unassigned_Secondary/Total,
                               Unassigned_NonSplit_pct = 100*Unassigned_NonSplit/Total,
                               Unassigned_NoFeatures_pct = 100*Unassigned_NoFeatures/Total,
                               Unassigned_Overlapping_Length_pct = 100*Unassigned_Overlapping_Length/Total,
                               Unassigned_Ambiguity_pct = 100*Unassigned_Ambiguity/Total
                               )]

fc.name.pct <- names(fc.assign)[names(fc.assign) %like% "_pct"]
fc.assign.tabl <- cbind(fc.assign[,1], fc.assign[, ..fc.name.pct])
fc.assign.tabl <- melt(fc.assign.tabl, id.vars = c("Sample"))
fc.assign.tabl <- fc.assign.tabl[, variable := gsub("_pct", "", variable)]
fc.assign.tabl <- fc.assign.tabl[, variable := factor(variable, levels = c("Assigned","Unassigned_Unmapped","Unassigned_MappingQuality","Unassigned_Chimera","Unassigned_FragmentLength","Unassigned_Duplicate","Unassigned_MultiMapping","Unassigned_Secondary","Unassigned_NonSplit","Unassigned_NoFeatures","Unassigned_Overlapping_Length","Unassigned_Ambiguity"))]

fc.name.not0 <- fc.assign.tabl[, .(sum = sum(value)), by = variable][sum > 0,]$variable
fc.assign.tabl <- fc.assign.tabl[variable %in% fc.name.not0,]

plot.fc.assign <- ggplot(fc.assign.tabl, aes(Sample, value, fill = variable)) +
  geom_bar(stat = "identity" , width = 0.8, position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = rseqc.palette, name = "") +
  coord_flip() +
  scale_y_continuous(name = "Percentages", breaks = seq(0,100,10)) +
  xlab("") +
  scale_x_discrete(limits=rev) +
  theme_bw() +
  theme(legend.position="bottom") +
  theme(legend.key.size = unit(0.6,"line")) +
  guides(fill=guide_legend(nrow=3, byrow=TRUE)) +
  theme(plot.title = element_text(face="bold")) +
  ggtitle("featureCounts: Assignments")

```

- Read count assignment from **featureCounts** for RNA-Seq is quite high (between `r fc.assign.min.perc`% and `r fc.assign.max.perc`%, calculated only from uniquely mapped reads)
  - Number of assigned reads is above limits of the recommended number of reads (~`r fc.assign.min`-`r fc.assign.max`M).
  - Read assignment rates are summarized in Table \@ref(tab:fc-assign-tab) and `r paste0("Figure \\@ref(fig:fc-assign)")`.

`r if(var.rsem==1){paste0("- Read count assignment from **RSEM** for RNA-Seq is very high (between ",rsem.assign.min.perc," % and ",rsem.assign.max.perc," %).")}`
`r if(var.rsem==1){paste0("  - Number of assigned reads is on a border of the recommended number of reads (~",rsem.assign.min,"-",rsem.assign.max,"M).")}`
`r if(var.rsem==1){paste0("  - Read assignment rates are summarized in Table \\@ref(tab:rsem-assign-tab) and Figure \\@ref(fig:rsem-assign)")}`

```{r tab.id="fc-assign-tab", tab.cap="Summary of read-to-gene assignment. Percentages counted only from uniquely mapped reads - featureCounts.", results='asis', echo=FALSE, ft.split=F}
fit_to_width(autofit(qflextable(fc.assign[,.(Sample, Assigned, "Assigned (%)"= round(percent_assigned, 2))])),max_width = 9.1)
```

```{r fc-assign, results='asis', echo=FALSE, fig.cap="Read count assignment to genes (featureCounts).", fig.align="center", out.width="90%", fig.show="hold", dev="svg", fig.width=7, fig.height = custom_height1}
plot.fc.assign
```


```{r tab.id="rsem-assign-tab", tab.cap="Summary of read-to-gene assignment. Percentages counted only from uniquely mapped reads - RSEM.", results='asis', echo=FALSE, ft.split=F}
if(var.rsem==1){
fit_to_width(autofit(qflextable(rsem.assign[,.(Sample, Alignable, "Alignable (%)"=round(alignable_percent,2))])),max_width = 9.1)
}
```

```{r rsem-assign, results='asis', echo=FALSE, fig.cap="Read count assignment to genes (RSEM).", fig.align="center", out.width="90%", fig.show="hold", dev="svg", fig.width=7, fig.height = custom_height1}
if(var.rsem==1){
  plot.rsem.assign
}
```


### Biotypes
- All the samples captured mainly protein-coding genes (Figure \@ref(fig:biotype)).
  - This is expected from polyA selection based libraries.

```{r biotype, fig.cap = "Captured gene biotypes.", echo=FALSE, out.width="90%", fig.width=7, fig.height = custom_height1, fig.align="center", fig.show="hold", dev="svg"}
biotype.files <- list.files(paste0(INPUTFILES), pattern = ".biotype_counts.txt", recursive = T)
featureCounts <- fread(paste0(INPUTFILES,biotype.files[1]), header=T, stringsAsFactors = F)
if (length(biotype.files) > 1) {
  for (i in 2:length(biotype.files)){
    sample <- fread(paste0(INPUTFILES,biotype.files[i]), header=T, stringsAsFactors = F)
    featureCounts <- merge(featureCounts,sample,by = "Geneid")
  }
}
colnames(featureCounts)[-1]<-sapply(biotype.files,function(x) gsub(".biotype_counts.txt","",basename(x)))  # Rename columns - split after first "_"

featureCounts <- melt(featureCounts, id.vars = "Geneid", variable.name = "Sample")
featureCounts <- merge(config_json[, .(Sample = full_name)], featureCounts, by = "Sample")

featureCounts[, perc := round((100*value/sum(value)),2), by=Sample] # compute percentage
featureCounts[, sum.value := sum(value), by = Sample]
pct <- featureCounts[perc >=2 ,] # Get abundant features (>= 2%)
others <- pct[, .(Geneid = "other", value = sum.value - sum(value), perc = round(100-(sum(perc)),2)), by = .(Sample, sum.value)] # Other features - added together to one
others <- others[, .(Geneid, Sample, value, perc, sum.value)]

pct <- rbind(pct, others)

setorder(pct, Sample, Geneid)

### Plot Pie plot only if one sample
if(length(unique(pct$Sample)) == 1){
  #pdf(file=paste0(outpdf), width = 6, height = 6)

  piecolor <- brewer.pal(n = length(pct$Geneid), name = "Set1")
  pie(pct$perc, labels = paste0(pct$perc,"%"),
    col = piecolor,
    main = paste("Pie Chart of Gene Biotypes", pct$Sample[1], pct$sum.value[1], sep="\n"),
    cex=1)
  legend("bottomright", pct$Geneid, cex=0.7, fill=piecolor, title = "Biotype")

  #dev.off()
}else{
  mybar <- ggplot(data = pct, aes(x=Sample, y=perc, fill=Geneid)) +
          geom_bar(stat="identity") +
          scale_fill_brewer(palette = "Set1") +
          theme_bw() +
          coord_flip() +
          scale_y_continuous(name = "Percentages", breaks = seq(0,100,20)) +
          xlab("") +
          scale_x_discrete(limits=rev) +
          theme(plot.title = element_text(face="bold")) +
          ggtitle("Gene Biotypes") +
          theme(legend.position="bottom") +
          theme(legend.key.size = unit(0.6,"line")) +
          guides(fill=guide_legend(nrow=2,byrow=TRUE))
  #ggsave(filename = paste0(outpdf), plot = mybar, device = pdf, width = 10, height = 6)
  mybar
    }
```


# Notes and comments
- We strongly recommend confirming all the events manually in genome/alignment browsers such as IGV, Tablet or Savant.
  - All the alignments and references are provided.

## Visualization

### IGV
- Reference genomes and annotations for `r reference` are available to download on [Ensembl](https://www.ensembl.org/index.html) website.
- You can upload the provided aligned bam files with the reference in IGV for visualization.
- The IGV tutorial is available [here](http://software.broadinstitute.org/software/igv/userguide).
- All the index files (.bai, .fai, .idx) must be in the same folder as their "parent" files otherwise IGV wouldn't open them.

# DE analysis
- For the DE analysis we used gene counts from `r ifelse(var.rsem==1,"RSEM","featureCounts")` tool.

## Expression-based quality check
- Read counts for coding genes (mostly protein-coding genes) distribution is depicted in Figure \@ref(fig:de-norm-count).
- Post-normalization counts (~normalized expressions) seem to be OK (Figure \@ref(fig:de-norm-count)).
- Gene expression dispersion look normal.

```{r de-norm-count, fig.cap = paste0("Number of reads assigned to coding genes for all the samples before normalization (raw counts). Coding genes depth for all the samples after normalization (normalized counts)."), echo=FALSE, out.width="80%", fig.align="center", results="asis", fig.show="hold", dev="svg"}

if(config_json$conditions_to_compare[1] == "all"){
  ppnc.files <- list.files(path=paste0("../results/",DE,"/",comparison[1],"/all/"),pattern ="pre_post_norm_counts.svg", full.names = T)
}else{
  ppnc.files <- list.files(path=paste0("../results/",DE,"/",comparison,"/all/"),pattern ="pre_post_norm_counts.svg", full.names = T)
}

if(length(ppnc.files)>1){
  for(i in 2:length(ppnc.files)){
    cat("<div align='center'><img src=",ppnc.files[i]," width='80%'/></div>")
    #cat("![](",ppnc.files[i],")")
    cat('\n\n')
  }
}

knitr::include_graphics(ppnc.files[1])
```

- Clustering of the samples (based on genes expression) shows that all the samples cluster together by their condition.
  - This is summarized in heatmap in Figure \@ref(fig:de-heatmap).
- PCA shows a similar effect as the previous clustering.
  - PCA is depicted in Figure \@ref(fig:de-pca).

```{r de-heatmap, fig.cap = "Clustering of samples by the gene expression (DESeq2 log2 normalization).", echo=FALSE, out.width="80%", fig.align="center", results="asis", fig.show="hold", dev="svg"}

if(config_json$conditions_to_compare[1] == "all"){
  heatmap.sample.files <- data.table(comp = comparison[1])
  heatmap.sample.files[, path:=ifelse(file.exists(paste0("../results/",DE,"/",comp,"/all/heatmaps_samples_log_batch.png")), paste0("../results/",DE,"/",comp,"/all/heatmaps_samples_log_batch.png"), paste0("../results/",DE,"/",comp,"/all/heatmaps_samples_log.png"))]
}else{
  heatmap.sample.files <- data.table(comp = comparison)
  heatmap.sample.files[, path:=ifelse(file.exists(paste0("../results/",DE,"/",comp,"/all/heatmaps_samples_log_batch.png")), paste0("../results/",DE,"/",comparison,"/all/heatmaps_samples_log_batch.png"), paste0("../results/",DE,"/",comp,"/all/heatmaps_samples_log.png"))]
}

#heatmap.sample.files = list.files(paste0(mRNA_DE_FOLDER_PATH,comparison$cp,"/all/"), pattern = "heatmaps_samples_log", full.names = T)

if(length(heatmap.sample.files$path)>1){
  for(i in 2:length(heatmap.sample.files$path)){
    cat("<div align='center'><img src=",heatmap.sample.files$path[i]," width='80%'/></div>")
    #cat("![](",heatmap.sample.files$path[i],")")
    cat('\n\n')
  }
}

knitr::include_graphics(heatmap.sample.files$path[1])
```


```{r de-pca, fig.cap = "PCA (first two components) visualization of all samples (DESeq2 VST normalization).", echo=FALSE, out.width="80%", fig.align="center", results="asis", fig.show="hold", dev="svg"}

if(config_json$conditions_to_compare[1] == "all"){
  sample.sample.files <- data.table(comp = comparison[1])
  sample.sample.files[, path:=ifelse(file.exists(paste0("../results/",DE,"/",comp,"/all/sample_to_sample_PCA_batch.svg")), paste0("../results/",DE,"/",comp,"/all/sample_to_sample_PCA_batch.svg"), paste0("../results/",DE,"/",comp,"/all/sample_to_sample_PCA.svg"))]
}else{
  sample.sample.files <- data.table(comp = comparison)
  sample.sample.files[, path:=ifelse(file.exists(paste0("../results/",DE,"/",comp,"/all/sample_to_sample_PCA_batch.svg")), paste0("../results/",DE,"/",comp,"/all/sample_to_sample_PCA_batch.svg"), paste0("../results/",DE,"/",comp,"/all/sample_to_sample_PCA.svg"))]
}

#sample.sample.files = list.files(paste0(mRNA_DE_FOLDER_PATH,comparison$cp,"/all/"), pattern = "sample_to_sample_PCA.png", full.names = T)

if(length(sample.sample.files$path)>1){
  for(i in 2:length(sample.sample.files$path)){
    cat("<div align='center'><img src=",sample.sample.files$path[i]," width='80%'/></div>")
    #cat("![](",sample.sample.files$path[i],")")
    cat('\n\n')
  }
}

knitr::include_graphics(sample.sample.files$path[1])
```

## Differential expression results

- Please note that for visualization purposes we had to set a predefined cut-off value for some of the parameters (adj.p-value and logFC)
  - **Please read the disclaimer at the beginning of the report for more information about selected cut-off values!**

- In this section, we might provide several results as the number of compared conditions is variable between experiments
- Differential gene expression was calculated by two separate tools - edgeR and DESeq2
  - Both of the tools have different performances: edgeR is usually more sensitive and less specific, DESeq2 is usually more specific and less sensitive; edgeR is more suitable for fewer replicates (<12) as it is not as conservative as DESeq2
  - Generally, DESeq2 is more conservative than edgeR.
  - If your goal is an exploratory analysis with a low number of replicates and with low expressed genes, edgeR might be your choice.
  - If you need a selection of genes which are strongly differentially expressed and you want to remove as many false-positive results as possible for a price of some false-negative results, DESeq2 might be your choice.
  - If you want to perform some sort of prioritization, you might select an overlap between the two tools.
  - **Note for DESeq2**: By default, DESeq2 applies independent filtering which aims at removing genes which are potential outliers or show “strange” behavior (~large variance). Filtered genes are then marked by NA in either p-value or adj.p-value column. This filtering might be in some cases too strict and might cause a loss of interesting results. For this reason, we also make DESeq2 results without the independent filtering (DESeq2_noIndFilt). If you choose DESeq2 results, I recommend to start with filtered results but look at the end of the table at the filtered genes (genes with NA in p-value/adj.-value are at the bottom of the table). In case you see some interesting results there you might consider switching to the unfiltered results. In edgeR analysis, only genes with very low expression are excluded (1 read-per-million reads in at least 3 samples).
- The full description of a DE results is given after in the Output files section

- Number of DE genes with default cut-off values is summarized in Table \@ref(tab:tabl-de-de)
  - Please note this is only a tentative view on approximate differences between the conditions and not final results.
- Volcano plot visualization visualizes log2FC (x-axis) and adj.p-value (y-axis).
  - Colored dots highlight genes above the defined adj.p-value cut-off value, blue lines show defined log2FC cut-off value.
  - Volcano plot from DESeq2 results can be found in Figure \@ref(fig:de-volcano).
- MA plot visualizes mean expression (x-axis) and log2FC (y-axis).
  - Colored dots highlight genes above the defined cut-off values (both adj.p-value and log2FC).
  - MA plot from DESeq2 results can be found in Figure \@ref(fig:de-ma-plot).
- An example of maximum top 20 most DE genes (by adj.p-value) can be found in Figure \@ref(fig:de-top-de).
  - The selection of a maximum 20 genes is always based on compared conditions but visualizes expression values from all other conditions if applicable.

```{r plot-capt, results="asis", echo=FALSE}
capt <- paste0("Summary or number of DE genes based on default cut-off values from all three DE calculations. The first two columns are counts with both adj.p-value and logFC cut-off, the other two columns are only with adj.p-value cut-off. DESeq2 results, DESeq2 results without independent filtering (noIndFilt) and edgeR results are shown.")
```


```{r tab.id="tabl-de-de", tab.cap=capt, echo=FALSE, results="asis", ft.split=F}
tabl.de.de.all <- data.table()

for(i in 1:length(comparison)){
  de.tab.edger <- fread(paste0("../results/",DE,"/",comparison[i],"/all/edgeR.tsv"),sep  = "\t",header = T)
  de.tab.deseq <- fread(paste0("../results/",DE,"/",comparison[i],"/all/DESeq2.tsv"),sep  = "\t",header = T)
  de.tab.deseq.noIndFilt <- fread(paste0("../results/",DE,"/",comparison[i],"/all/DESeq2_noIndFilt.tsv"),sep  = "\t",header = T)

  de.tab.edger.upl <- dim(de.tab.edger[padj < 0.05 & log2FoldChange >= 1,])[1]
  de.tab.edger.dnl <- dim(de.tab.edger[padj < 0.05 & log2FoldChange <= -1,])[1]
  de.tab.edger.up <- dim(de.tab.edger[padj < 0.05 & log2FoldChange > 0,])[1]
  de.tab.edger.dn <- dim(de.tab.edger[padj < 0.05 & log2FoldChange < 0,])[1]

  de.tab.deseq.upl <- dim(de.tab.deseq[padj < 0.05 & log2FoldChange >= 1,])[1]
  de.tab.deseq.dnl <- dim(de.tab.deseq[padj < 0.05 & log2FoldChange <= -1,])[1]
  de.tab.deseq.up <- dim(de.tab.deseq[padj < 0.05 & log2FoldChange > 0,])[1]
  de.tab.deseq.dn <- dim(de.tab.deseq[padj < 0.05 & log2FoldChange < 0,])[1]

  de.tab.deseq.noIndFilt.upl <- dim(de.tab.deseq.noIndFilt[padj < 0.05 & log2FoldChange >= 1,])[1]
  de.tab.deseq.noIndFilt.dnl <- dim(de.tab.deseq.noIndFilt[padj < 0.05 & log2FoldChange <= -1,])[1]
  de.tab.deseq.noIndFilt.up <- dim(de.tab.deseq.noIndFilt[padj < 0.05 & log2FoldChange > 0,])[1]
  de.tab.deseq.noIndFilt.dn <- dim(de.tab.deseq.noIndFilt[padj < 0.05 & log2FoldChange < 0,])[1]

  tabl.de.de <- data.table(Analysis=c("DESeq2","DESeq2 noIndFilt","edgeR"),
                up_pval_logfc=c(de.tab.deseq.upl,de.tab.deseq.noIndFilt.upl,de.tab.edger.upl),
                down_pval_logfc=c(de.tab.deseq.dnl,de.tab.deseq.noIndFilt.dnl,de.tab.edger.dnl),
                up_pval=c(de.tab.deseq.up,de.tab.deseq.noIndFilt.up,de.tab.edger.up),
                down_pval=c(de.tab.deseq.dn,de.tab.deseq.noIndFilt.dn,de.tab.edger.dn))
  tabl.de.de <- tabl.de.de[,.(Comparison = comparison[i], Analysis, up_pval_logfc, down_pval_logfc, up_pval, down_pval)]

  tabl.de.de.all <- rbind(tabl.de.de.all,tabl.de.de)
}

#knitr::kable(x = tabl.de.de, booktabs = TRUE, caption = capt, format.args = list(big.mark = ","))
edgeR.tag.pos <- which(tabl.de.de.all == "edgeR", arr.ind = TRUE) # find row and column with "edgeR"
ftabl.de.de.all <- flextable(tabl.de.de.all[,.(Comparison, Analysis, "Up (adj.pval & logfc)"=up_pval_logfc, "Down (adj.pval & logfc)"=down_pval_logfc,"Up (adj.pval)"=up_pval,"Down (adj.pval)"=down_pval)]) # change the table to a flextable object

# apply conditional formating - for each row with "edgeR" add a horizontal line
for(i in 1:nrow(edgeR.tag.pos)) {
  ftabl.de.de.all <- ftabl.de.de.all %>%
    hline(i = edgeR.tag.pos[i, 1], border = fp_border(color = "black", style = "solid", width = 1))
}

fit_to_width(autofit(ftabl.de.de.all), max_width = 9.1)
```


```{r de-volcano, fig.cap = paste0("Volcano plot (DESeq2 results with independent filtering on). Maximum of 20 most DE genes (by adj.p-value) are named."), echo=FALSE, out.width="80%", fig.align="center", results="asis", fig.show="hold", dev="svg"}

volcano.files <- grep(grep(grep(list.files(path=paste0("../results/",DE,"/",comparison,"/all/"), pattern = "volcano", full.names = T), pattern="Filt", value = T, invert = T), pattern="edgeR", value = T, invert = T), pattern="ggplot2.png", value=T)

if(length(volcano.files)>1){
  for(i in 2:length(volcano.files)){
    cat("<div align='center'><img src=",volcano.files[i]," width='80%'/></div>")
    #cat("![](",volcano.files[i],")")
    cat('\n\n')
  }
}

knitr::include_graphics(volcano.files[1])
```


```{r de-ma-plot, fig.cap = paste0("MA plot (DESeq2 results with independent filtering on). Maximum of 20 most DE genes (by adj.p-value) are named."), echo=FALSE, out.width="80%", fig.align="center", results="asis", fig.show="hold", dev="svg"}

ma.files <- grep(grep(grep(list.files(path=paste0("../results/",DE,"/",comparison,"/all/"), pattern = "MAplot_", full.names = T), pattern="Filt", value = T, invert = T), pattern="edgeR", value = T, invert = T), pattern="ggpubr.png", value=T)

if(length(ma.files)>1){
  for(i in 2:length(ma.files)){
    cat("<div align='center'><img src=",ma.files[i]," width='80%'/></div>")
    #cat("![](",ma.files[i],")")
    cat('\n\n')
  }
}

knitr::include_graphics(ma.files[1])
```


```{r de-top-de, fig.cap = paste0("Heatmap of maximum top 20 differentially expressed genes (DESeq2 results with independent filtering, adj.p-value < 0.05, logFC >= +/-1)."), echo=FALSE, out.width="80%", fig.align="center", results="asis", fig.show="hold", dev="svg"}

top_de.files <- list.files(path=paste0("../results/",DE,"/",comparison,"/all/"),pattern ="heatmap_selected_orderBaseMeanCluster.png", full.names = T)

if(length(top_de.files)>0){
  if(length(top_de.files)>1){
    for(i in 2:length(top_de.files)){
      cat("<div align='center'><img src=",top_de.files[i]," width='80%'/></div>")
      #cat("![](",top_de.files[i],")")
      cat('\n\n')
    }
  }

knitr::include_graphics(top_de.files[1])
}
```

## Used statistics and selection of differentially expressed genes

### Log2FC and adjusted p-value
- Differences between the conditions are expressed in **log2FC** (log2 of fold-change)
    - This is calculated by comparing average expression one condition vs the other and then the log2 transformation
    - This makes the results symmetric around 0 and it’s easier to understand
    - For example: -1 is a 2-fold decrease of the expression, +1 is a 2-fold increase of the expression.
    - Please note it always depends on the “direction” of the comparison - if you are not sure just compare normalized counts of one gene to see from which “direction” the log2FC was calculated.
- Calculation of DE genes consists of a very high number of comparisons (thousands of genes = thousands of comparisons)
  - For this reason, we need to correct raw p-values for false discovery rate which emerges from the number of comparisons
  - This results in **adjusted p-values** (adj.pval, adj. p-values, adj. pval)
    - In case you want to use some filtering or selection based on the statistical evidence you should ALWAYS USE adjusted p-values and never raw p-values
    - Raw p-values can be used only in very special cases and always require a discussion with bioinformatician or statistician

## Note on selecting of differentially expressed genes (DEG)
- Selection of differentially expressed genes always has to be primarily based on the biological background of the experiment and the hypothesis
- A lot of people select DEG based on a combination of cut-off values
- This might be tricky as no predefined cut-off values exist and it depends on the experiment itself, on the amount of false-positive results you are willing to accept and on the minimum level of changes if any
- Commonly used statistical values, such as logFC and adj.p-value, should help you to confirm the hypothesis, results or the observations and not lead the discovery itself
- If you decide to go for some cut-off values, do not blindly select only genes above the thresholds!
  - For example, the frequently used cut-off value of adj.p-value 0.05 at which you will reject the hypothesis the gene is not expressed the same: one gene will have adj.p-value 0.0499 which would pass this cut-off whereas the second gene will have adj.p-value of 0.0501 and would not pass the cut-off.  The difference between the genes is negligible and still, the first gene would and the other gene would not be accepted if a strict cut-off value would be set.
  - Another “issue” with adj.p-value is when the groups we are comparing are heterogeneous (for example clinical samples). In this case, adj.p-value are generally much higher.  In this situation, we should strongly focus on biological effect and hypothesis and use statistics to either confirm the observation or to function for prioritization of the results
  - The similar thing applies to logFC
  - In the case of logFC, we also have to consider the effect of the expression of individual genes. For example, the frequently used cut-off value of 1 (2x fold-change): A gene with expression 10 in one group and 5 in other group results in logFC 1 and would pass the cut-off. The overall expression of the gene is negligible and most likely the change doesn’t have any biological effect. Another gene has an expression of 10000 in one group and 5001 in another group. This gene would not pass the cut-off value but most likely the change will have some kind of biological effect.
  - Sometimes small logFC might be more important than high logFC and the same applies to adj.p-value.
- We should always choose the genes based on the overall properties of the gene and the comparison between conditions

## Output files
- Results are organized by compared conditions
  - The compared conditions are always stated in the name of the folder
  - All the results in the corresponding folder are based on the currently compared conditions but in some of the cases other samples and conditions are added to the visualization/summaries as well
  - For example: **cond1_vs_cond2** compares cond1 to cond2. In the DE results then - positive log2FC signalizes more expression in cond1 (or less in cond2) and negative log2FC signalizes less expression in cond2 (or more in cond1).
- Main output files are **DESeq2.tsv**, **DESeq2_noIndFilt.tsv**, and **edgeR.tsv**
- All three main results contain several columns:
  - **first column (no name)** - Ensembl gene id
  - **baseMean** - an average expression of genes from samples from compared conditions
  - **log2FoldChange** - log2FC of the gene expression difference. The difference is calculated by comparing average expression of samples of one condition with the other samples from the other condition
  - **stat and lfcSE** (in DESeq2); LR and tgw.Disp (in edgeR) - results of the statistical test performed by the tool and variation/dispersion of the gene
  - **pvalue** - a raw p-value of the comparison
  - **padj** - multiple testing correction (Benjamini-Hochberg correction) p-value of the comparison. In case you are comparing and analyzing RNA-Seq use this as the evaluation of the statistical significance of the comparison
  - **gene_name** - common gene name
  - **gene_biotype** - biotype of the gene
  - **_normCounts** - normalized gene expression for each sample (not only the samples in the current comparison)
  - **_rawCounts** - raw read counts for each sample (not only the samples in the current comparison)
- The full description is given below

- Additional visualizations and/or a different combination of visualization is available upon request after discussion

## General/common outputs
- **xxx** in the name of some pdf files in the description below represents compared conditions
- **background.txt** – list of genes that were expressed in your samples (=gene was expressed in at least one of the samples). The first column is Ensembl gene ID, second is common gene name.
- **norm_counts.xlsx** - DESeq2 normalized counts (~expression) for all genes and all samples. Can be used for visualization of the data.
  - **norm_counts.tsv** - the same information as in norm_counts.xlsx but in a text form.
- **all_sig_genes_normCounts.pdf** – significantly DEG (here, “default cut-off” adj.p-value 0.05 and logFC 1 is considered) expression visualization. Selection of DE genes is based on the particular comparison but all conditions/samples are plotted. This is based on DESeq2 results.
- **counts_barplot.pdf** - visualization of used (raw) read counts per sample
- **DESeq2.tsv** – main results from differential gene expression by DESeq2. Description of the columns is given above. tsv is a tab-separated source file for the xlsx.
- **DESeq2_noIndFilt.tsv** – results from differential gene expression without independent gene filtering (see Independent filtering and Cooks cut-off in DESeq2 manual). tsv is a tab-separated source file for the xlsx.
- **DESeq2_de_genes_check.txt** and **DESeq2_de_genes_check_noIndFilt.txt** – basic summary of differential expression. “Default” adj.p-value of 0.05 and logFC of 1 are used. The upper part of the table considers only adj.p-value cut-off value, the lower part considers adj.p-value and logFC. This servers only for a  demonstration of the possible effect, not as a final result
- **DESeq2_design_control.txt** - a control file for used sample design
- **edgeR.tsv** – main results from differential gene expression by edgeR. tsv is a tab-separated source file for the xlsx.
- **edgeR_de_genes_check.txt** - same as in case of DESeq2 - “quick and dirty” check of the effect to get a general idea about the differences.
- **heatmap_selected_orderBaseMean.pdf** – heatmap of maximum top 20 most differentially expressed genes (based on adj. p-value) ordered by an average expression. Blue means low expression, red means high expression. The maximum top 20 genes are selected based on the compared groups but all samples are visualized. This is based on DESeq2 results.
- **heatmap_selected_orderBaseMeanCluster.pdf** – heatmap of maximum top 20 most differentially expressed genes (based on adj. P-value, same as above) clustered by rows and columns. Blue means low expression, red means high expression. This is based on DESeq2 results.
- **heatmaps_samples.pdf** - heatmap visualization of sample clustering based on gene expression. Several normalizations are applied but the first one usually looks the best.
- **MAplot_xxx_vs_xxx_ggpubr.pdf** - MA plots for the gene expression. Maximum top 20 most DEG (by adj.p-value) are named. This is based on DESeq2 results.
- **MAplot_xxx_vs_xxx_noIndFilt_ggpubr.pdf** -  the same as above but without independent filtering.
- **MDS_plot.pdf** - MDS-based visualization of sample clustering by expression
- **overlap_DESeq2_edgeR_de.tsv** and **overlap_DESeq2_edgeR_venn.pdf** – overlap between DE genes between DESeq2 and edgeR calculation. In default, genes with adj. p-value <0.05 and logFC +/- 1 are included. Columns DESeq2 and edgeR contain values TRUE or FALSE and this represents whether the resulting gene was included in the results by an individual tool.
- **pre_post_norm_counts.pdf** - visualization of used raw (top figure) and normalized (bottom) read counts per sample
- **sample_to_sample_PCA.pdf** - PCA-based visualization of sample clustering by expression. The first figure shows first two PCA components, the second shows all combinations of first three components (~3D look)
- **volcanoplot_xxx_vs_xxx_ggplot2.pdf** – volcano plot of the results from DESeq2. Maximum top 20 most DEG (based on adj.p-value) are depicted. This is based on DESeq2.
- **volcanoplot_xxx_vs_xxx_noIndFilt_ggplot2.pdf** - same as above but without independent filtering.

- Please note the differences between MA and Volcano plots - different X and Y axes giving you a different view on the results.

- Alignments can be visualized in genome browsers such as [IGV](http://software.broadinstitute.org/software/igv/), [Tablet](https://ics.hutton.ac.uk/tablet/) or [Savant](http://www.genomesavant.com/p/savant/index)
  - For more details please read the manual or contact your bioinformatician
- If the mapping files were shared through the online storage please download them as soon as possible and let us know so we can delete them

# Output files (alphabetical order)
## Folders
### Raw_fastq
- Provided raw sequenced fastq files

### Alignment
- Provided alignments in bam format

### QC_general
- html file summarizing most of the QC done

### FeatureCounts_expression
- Gene expression raw counts for each sample

### RSEM_expression
- Gene and transcript expression for each sample
- The normalized expressions are summarized in TPM column

#### Files
- .genes.results.tsv - gene expression estimates
- .isoforms.results.tsv - transcript expression estimates

### UCSC_signals
- Alignment signal for UCSC browser
- Two signal “types” per sample are provided:
  - Unique - signal only from uniquely mapped reads
  - UniqueMultiple - signal from both uniquely and multi-mapped reads
- Each set is then split into two:
  - str1 - signal coming from the alignment to the plus strand of DNA
  - str2 - signal coming from the alignment to the minus strand of DNA
- If you have sense/forward specific library, the signal strand corresponds to the strand of the annotated gene (mapping ----> corresponds to the [+] strand annotation)
- If you have antisense/reverse specific library, the signal strand corresponds to the opposite strand gene annotation (mapping <---- corresponds to the [+] strand annotation)

### DE_analysis
- All files related to Differentially expressed genes analysis

## Gene ontology, GSEA and pathway analysis
- This section summarizes possible follow-up analyses/interpretations of the differential expression results
- In the following analyses the input can sometimes be specified as Ensembl ID (usually provided in the results unless noted otherwise), some tools required common gene names (usually provided in the results as well) or Entrez ID (NCBI/RefSeq)
  - This is always specified in the manual of the tool
- GO and/or Pathway analysis can be performed for up-/down-regulated genes as well as for the whole list of de-regulated genes
  - Both of the analyses give a different view of the results
    - Analysis of separate up-/down-regulated genes gives you a direct answer on the change, in particular, GO or Pathway and the “direction” of the regulation
    - Analysis of all de-regulated genes (up-/down-regulated together) gives you a more broad view of the total change
  - Both of the results are helpful but you have to realize what kind of information you can get and what kind of answer they provide

### Gene Ontology
- Input for gene ontology (GO) testing is usually a list of differentially expressed genes
  - You can either use filtering by some cut-off value(s) or select top X genes sorted by adj.p-value/logFC/...
- You can either analyze all DEG or separately down-/up-regulated
  - Please see the section above for more details
- A second input is gene background/universe
  - This is a list of all genes expressed in the comparison
  - This list is used as a “reference” for the statistical evaluation of significantly changed GO groups
- For “common” organisms like human or mouse, a very nice tool is [GOrilla](http://cbl-gorilla.cs.technion.ac.il/) with very nice figures and direct export to Excel
  - It also contains links to [REViGO](http://revigo.irb.hr/) which can help you to summarize the GO groups into more general categories
- To simply browse GO categories you can use [QuickGO](http://www.ebi.ac.uk/QuickGO/) or [Ontology Lookup Service](http://www.ebi.ac.uk/ols/index)
- Another option to filter down GO categories is [AmiGO2](http://amigo.geneontology.org/amigo/dd_browse).
- If, for whatever reason, you don't have gene ontology annotations but just sequences of your genes you can still get GO analysis. [Blast2GO](https://www.blast2go.com/) is a nice service combining BLAST similarity search with GO annotation to produce GO analysis. There is a free and paid version where the free version gives you basic annotations and have some pretty visualizations. A tutorial is for example here 	https://www.youtube.com/watch?v=GqSqS_izlYg&t=335s.
- Other tools include [g:Profiler](https://biit.cs.ut.ee/gprofiler/) and many more

### Pathway analysis
- Input data are very similar to GO (mentioned above)
- [PANTHER](http://www.pantherdb.org ) and [Reactome](http://www.reactome.org/) provide quite nice analysis for the Pathway analysis and exploration.
  - PANTHER recently started to include GO testing as well so it became a multi-purpose tool
- [KEGG](http://www.genome.jp/kegg/kegg2.html) is another commonly used tools for pathway analysis but it is not that easy to interpret the data unless you know which pathways should be involved

### Gene Set  Enrichment
- Input data to Gene Set Enrichment are a bit different than in previous cases
- You do NOT subsample your results but you take a whole list of gene expression results and sort it according to specified criteria (logFC, adj.p-value, ...)
  - In the case of logFC, you have the most positively changed genes on the top and the most negatively changed genes at the bottom and non-changed genes are in the middle but they are still in the SAME list
  - The enrichment then compares different parts of the list and looks for patterns or similarities
- One of the most used is probably [GSEA](http://software.broadinstitute.org/gsea/index.jsp)
  - It’s a standalone tool which takes some time to learn but provides reasonable results
- [GOrilla](http://cbl-gorilla.cs.technion.ac.il/) also offers gene enrichment analysis
- Another option is to use [DAVID](https://david.ncifcrf.gov/home.jsp) which is also a multipurpose tool but I do not recommend it too much since the interpretation of the results might be difficult and very subjective. But it can still be used if there is no other option or you find it easy to work with.

### Protein-protein interactions
- [HIPPIE](http://cbdm-01.zdv.uni-mainz.de/~mschaefer/hippie/index.php) provides reliable and meaningful human protein-protein interaction networks. It is suitable more for gene-by-gene exploration.
- [STRING](https://string-db.org/cgi/input.pl?UserId=4hzU6jHtBui6&sessionId=fokvQzUPEo8M&input_page_show_search=off) is a database of known and predicted protein-protein interactions supported by SIB, CPR - NNF and EMBL. This includes physical, functional as well as predicted interactions.
- [BioGRID](https://thebiogrid.org/) is an interaction repository with data compiled through comprehensive curation efforts. Contains interactions extracted from publications for major model organisms

### Multipurpose tools and other
- There are other tools/approaches to help with the secondary analysis
- [Enrichr](http://amp.pharm.mssm.edu/Enrichr/) is nice and simple to run the tool to identify enriched pathways, different gene summaries, and many others. It provides many various plots and tables
- [GeneMANIA](http://genemania.org/) visualizes genes and their interaction, co-expression as well as Gene Ontology, etc. It is suitable more for gene-by-gene exploration.
- [ConsensusPathDB-human](http://cpdb.molgen.mpg.de/) integrates several different approaches including pathway analysis
- [BioCyc Database Collection](https://biocyc.org/) is a collection of 9387 Pathway/Genome Databases (PGDBs), plus software tools for understanding their data.
- [KEGG database](http://www.genome.jp/kegg/kegg2.html) is a database resource that integrates genomic, chemical and systemic functional information. In particular, gene catalogs from completely sequenced genomes are linked to higher-level systemic functions of the cell, the organism, and the ecosystem. KEGG accepts common gene names and NCBI gene/transcript ID and UniProt ID. For the analysis, you can use http://www.genome.jp/kegg/tool/map_pathway2.html or http://www.genome.jp/kegg/tool/map_pathway1.html. KEGG looks for genes in your list and tries to highlight the pathway where selected genes have some kind of association.
- [Venny](http://bioinfogp.cnb.csic.es/tools/venny/) is a nice tool for a simple visualization of overlaps between up to three sets of data.

### R/Bioconductor
- If you know how to use **R** you can check [**clusterProfiler**](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) which provides many different analyses types and has very nice manual and tutorials
- [**topGO**](http://bioconductor.org/packages/release/bioc/html/topGO.html) is another R package which offers both gene ontology and gene set enrichment analysis
- [**ReactomePA**](https://bioconductor.org/packages/release/bioc/html/ReactomePA.html) allows for analysis of pathways with a slightly different source than KEGG
- To add GO categories (and other interesting information) to your list of genes, you can try [**biomaRt**](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) which can directly load the Ensembl database

